src/adaptors/multi_product.rs

```diff
@@ -171,81 +102,114 @@ where
     type Item = Vec<I::Item>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if MultiProduct::iterate_last(&mut self.0, MultiProductIterState::StartOfIter) {
-            Some(self.curr_iterator())
-        } else {
-            None
+        let inner = self.0.as_mut()?;
+        match &mut inner.cur {
+            Some(values) => {
+                for (iter, item) in inner.iters.iter_mut().zip(values.iter_mut()).rev() {
+                    if let Some(new) = iter.iter.next() {
+                        *item = new;
+                        return Some(values.clone());
+                    } else {
+                        iter.iter = iter.iter_orig.clone();
+                        // `cur` is not none so the untouched `iter_orig` can not be empty.
+                        *item = iter.iter.next().unwrap();
+                    }
+                }
+                // The iterator ends.
+                self.0 = None;
+                None
+            }
+            // Only the first time.
+            None => {
+                let next: Option<Vec<_>> = inner.iters.iter_mut().map(|i| i.iter.next()).collect();
+                if next.is_some() {
+                    inner.cur = next.clone();
+                } else {
+                    self.0 = None;
+                }
+                next
+            }
         }
     }
 
     fn count(self) -> usize {
-        if self.0.is_empty() {
-            return 0;
-        }
-
-        if !self.in_progress() {
-            return self
-                .0
-                .into_iter()
-                .fold(1, |acc, multi_iter| acc * multi_iter.iter.count());
+        match self.0 {
+            None => 0,
+            Some(MultiProductInner { iters, cur }) => {
+                if cur.is_none() {
+                    iters
+                        .into_iter()
+                        .map(|iter| iter.iter_orig.count())
+                        .try_fold(1, |product, count| {
+                            if count == 0 {
+                                None
+                            } else {
+                                Some(product * count)
+                            }
+                        })
+                        .unwrap_or_default()
+                } else {
+                    iters.into_iter().fold(0, |mut acc, iter| {
+                        if acc != 0 {
+                            acc *= iter.iter_orig.count();
+                        }
+                        acc + iter.iter.count()
+                    })
+                }
+            }
```

Comment:
This is rather complicated logic, and if you have the time to provide comments for it (and the other specialized methods), that would be much appreciated. I'd love to see `multi_product.rs` get to the point where it's obvious how everything fits together. I'd suggest some comments here, myself, but I don't have a great understanding of how this works. I don't think there are many other places in itertools where our methods peek through three or four layers of nested types. 

This *is* a breaking change, so we can be a bit more patient in merging it.

Reply:
Indeed, the `try_fold(...).unwrap_or...` is basically an optimized `.product()` and deserves a comment. I'm gonna add some comments where I think it's valuable.